etna/whatson/models.py

```diff
@@ -0,0 +1,633 @@
+                FieldPanel("contact_info"),
+            ],
+            heading=_("Booking information"),
+        ),
+    ]
+
+    def clean(self):
+        """
+        Check that the venue address and video conference information are
+        provided for the correct venue type.
+        """
+
+        if self.venue_type:
+            if self.venue_type == VenueType.HYBRID and (
+                not self.venue_address or not self.video_conference_info
+            ):
+                raise ValidationError(
+                    {
+                        "venue_address": _(
+                            "The venue address is required for hybrid events."
+                        ),
+                        "venue_space_name": _(
+                            "The venue space name is required for hybrid events."
+                        ),
+                        "video_conference_info": _(
+                            "The video conference information is required for hybrid events."
+                        ),
+                    }
+                )
+            elif self.venue_type == VenueType.IN_PERSON and not self.venue_address:
+                raise ValidationError(
+                    {
+                        "venue_address": _(
+                            "The venue address is required for in person events."
+                        ),
+                        "venue_space_name": _(
+                            "The venue space name is required for hybrid events."
+                        ),
+                    }
+                )
+            elif self.venue_type == VenueType.ONLINE and not self.video_conference_info:
+                raise ValidationError(
+                    {
+                        "video_conference_info": _(
+                            "The video conference information is required for online events."
+                        ),
+                    }
+                )
+        return super().clean()
+
+    def with_content_json(self, content):
+        """
+        Overrides Page.with_content_json() to ensure page's `start_date` and `end_date`
+        value is always preserved between revisions.
+        """
+        obj = super().with_content_json(content)
+        obj.start_date = self.start_date
+        obj.end_date = self.end_date
+        return obj
+
+    def save(self, *args, **kwargs):
+        """
+        Set the event start date to the earliest session start date.
+        Set the event end date to the latest session end date.
+        """
+        if self.event_sessions.all():
+            self.start_date = (
+                self.event_sessions.all()
+                .order_by("session_start_date")
+                .first()
+                .session_start_date
+            )
+            self.end_date = (
+                self.event_sessions.all()
+                .order_by("session_end_date")
+                .last()
+                .session_end_date
+            )
```

Comment:
It's currently a little unclear to me whether `start_date` and `end_date` come from EventBrite (as mentioned elsewhere), or whether we always want to set them from the session data. If the latter, that's fine, but let's make sure there's no conflicting information further up.